File: backend/app/schemas/autotool.py
"""Schemas for the Autotool connection config (3rd publishing mode).

A singleton config (stored in app_settings under "autotool_config") holding the
target ImportPosts endpoint + the X-Api-Key used to authenticate to the
external Autotool proxy. The key is Fernet-encrypted at rest and never returned
to the client — the read schema only signals whether one is set.
"""
from datetime import datetime
from typing import Any

from pydantic import BaseModel


class AutotoolConfigRead(BaseModel):
    """Masked view: the API key is never echoed back."""

    target_url: str | None = None
    api_key_configured: bool = False


class AutotoolConfigUpdate(BaseModel):
    """Partial update.

    Field semantics mirror the backup S3 secret pattern:
      * omitted / None  → leave unchanged
      * "" (empty)      → clear
      * non-empty       → set
    """

    target_url: str | None = None
    api_key: str | None = None


class AutotoolTestResult(BaseModel):
    """Same shape domains use for Test connection."""

    ok: bool
    status_code: int | None = None
    detail: str
    elapsed_ms: int | None = None


# ----- shared tables + POST request preview -----


class AutotoolTableItem(BaseModel):
    id: int
    name: str
    autotool_token: str | None = None
    csv_path: str | None = None
    row_count: int = 0
    column_count: int = 0
    updated_at: datetime


class AutotoolTablesPage(BaseModel):
    items: list[AutotoolTableItem]
    total: int
    page: int
    page_size: int


class ColumnRef(BaseModel):
    id: int
    name: str


class AutotoolDomainRequest(BaseModel):
    """One ImportPosts POST — a single ``PAGE_SIZE``-row page for one site.

    A domain with more rows than the page size produces several of these, one
    per ``start`` offset. ``total`` is the domain's full row count (so the
    importer can tell when it has reached the last page); ``row_count`` is the
    rows in THIS page.
    """

    site: str
    file: str
    csv_path: str
    start: int
    total: int
    row_count: int
    body: dict[str, Any]


class AutotoolPostPreview(BaseModel):
    """The POST requests that would be sent to ImportPosts for one table.

    The table is split by its site column into ONE request per distinct domain
    (Autotool needs one file per site). ``site_column_id`` is auto-detected by
    default and remappable via the query param. Shared ``headers`` carry a
    masked X-Api-Key (the real key is never returned). Step 2 (actual sending)
    isn't built yet, so this is a faithful preview.
    """

    method: str = "POST"
    url: str | None = None
    headers: dict[str, str]
    columns: list[ColumnRef] = []
    site_column_id: int | None = None
    detected_site_column_id: int | None = None
    page_size: int = 50
    domain_count: int = 0
    page_count: int = 0
    total_rows_matched: int = 0
    table_row_count: int = 0
    requests: list[AutotoolDomainRequest] = []
    target_configured: bool = False
    api_key_configured: bool = False


# ----- send runs (background, with a progress page) -----


class AutotoolRunCreate(BaseModel):
    table_id: int
    site_column_id: int | None = None
    page_size: int | None = None


class AutotoolRunItemRead(BaseModel):
    """One ImportPosts POST within a run — a single (domain, page)."""

    id: int
    site: str
    start: int
    total: int
    status: str  # 'queued' | 'sending' | 'sent' | 'failed' | 'skipped'
    external_id: Any | None = None  # the proxy id for this item's site
    status_code: int | None = None
    detail: str | None = None
    response_snippet: str | None = None
    elapsed_ms: int | None = None
    created_at: datetime


class AutotoolRunRead(BaseModel):
    """A run, as shown in the list."""

    id: int
    table_id: int | None = None
    table_name: str
    target_url: str
    page_size: int
    status: str  # 'queued' | 'running' | 'cancelled' | 'done' | 'failed'
    total: int
    sent: int
    failed: int
    skipped: int = 0
    created_at: datetime
    started_at: datetime | None = None
    finished_at: datetime | None = None


class AutotoolRunsPage(BaseModel):
    items: list[AutotoolRunRead]
    total: int
    page: int
    page_size: int


class AutotoolRunDetail(AutotoolRunRead):
    """A run plus a page of its items (the progress page)."""

    site_column_id: int | None = None
    error: str | None = None
    items: list[AutotoolRunItemRead] = []
    items_total: int = 0
    items_page: int = 1
    items_page_size: int = 50

File: frontend/lib/autotool.ts
import { api } from "./api";

/** Masked config: the API key is never returned, only whether one is set. */
export interface AutotoolConfig {
  target_url: string | null;
  api_key_configured: boolean;
}

export interface AutotoolConfigUpdate {
  /** omit = unchanged, "" = clear, non-empty = set */
  target_url?: string | null;
  /** omit = unchanged, "" = clear, non-empty = set */
  api_key?: string | null;
}

export interface AutotoolTestResult {
  ok: boolean;
  status_code: number | null;
  detail: string;
  elapsed_ms: number | null;
}

export function getAutotoolConfig(): Promise<AutotoolConfig> {
  return api<AutotoolConfig>("/autotool/config");
}

export function saveAutotoolConfig(
  payload: AutotoolConfigUpdate,
): Promise<AutotoolConfig> {
  return api<AutotoolConfig>("/autotool/config", {
    method: "PUT",
    body: payload,
  });
}

export function testAutotoolConfig(): Promise<AutotoolTestResult> {
  return api<AutotoolTestResult>("/autotool/config/test", { method: "POST" });
}

// ----- shared tables + POST request preview -----

export interface AutotoolTableItem {
  id: number;
  name: string;
  /** Display name of the table's owner (creator); null if unset/deleted. */
  owner_name: string | null;
  autotool_token: string | null;
  csv_path: string | null;
  row_count: number;
  /** Count + names of the columns actually exposed to Autotool (the operator's
   *  selection, or all when none is set); column_count === columns.length. */
  column_count: number;
  columns: string[];
  updated_at: string;
}

export interface AutotoolTablesPage {
  items: AutotoolTableItem[];
  total: number;
  page: number;
  page_size: number;
}

export interface AutotoolDomainRequest {
  site: string;
  file: string;
  csv_path: string;
  /** 0-based row offset of this page within the domain. */
  start: number;
  /** the domain's full row count (across all pages). */
  total: number;
  /** rows in THIS page (<= page size). */
  row_count: number;
  body: {
    sites: string[];
    /** ACM does the row-split, so the proxy only needs the file. From the 2nd
     *  request, `id` (returned by the 1st) groups them into one import job. */
    data: { file: string; id?: string | number };
  };
}

export interface AutotoolPostPreview {
  method: string;
  url: string | null;
  headers: Record<string, string>;
  columns: { id: number; name: string }[];
  /** Required roles (domain/site, post_type, slug, status) not among the
   *  included columns; empty = ready to send. See missingRequiredColumns. */
  missing_required_columns: string[];
  /** Rows whose included `mode` column is "append" — Autotool appends the
   *  Content to existing WP content, so re-sending duplicates it. 0 = none. */
  append_row_count: number;
  /** This table already had a send run that delivered ≥1 item — a re-send would
   *  append a second time. Escalates the append warning. */
  previously_sent: boolean;
  site_column_id: number | null;
  detected_site_column_id: number | null;
  /** effective rows-per-request used to page the table (clamped server-side). */
  page_size: number;
  domain_count: number;
  /** total POSTs = sum over domains of ceil(rows / page size). */
  page_count: number;
  total_rows_matched: number;
  table_row_count: number;
  requests: AutotoolDomainRequest[];
  target_configured: boolean;
  api_key_configured: boolean;
}

/**
 * Columns Autotool's WordPress importer requires in every CSV, matched by
 * EXACT, case-insensitive header name (the `domain` role also accepts `site`).
 * Mirrors REQUIRED_AUTOTOOL_COLUMNS in backend/app/services/autotool_files.py —
 * keep the two lists in sync.
 */
export const AUTOTOOL_REQUIRED_COLUMNS: { label: string; names: string[] }[] = [
  { label: "domain", names: ["domain", "site"] },
  { label: "post_type", names: ["post_type"] },
  { label: "slug", names: ["slug"] },
  { label: "status", names: ["status"] },
];

/** Required Autotool roles NOT covered by `columnNames` (exact, case-insensitive,
 *  trimmed). Pass the names that will actually be in the CSV — i.e. the checked
 *  subset — so an unchecked required column is reported missing. */
export function missingRequiredColumns(columnNames: string[]): string[] {
  const present = new Set(columnNames.map((n) => n.trim().toLowerCase()));
  return AUTOTOOL_REQUIRED_COLUMNS.filter(
    (role) => !role.names.some((n) => present.has(n)),
  ).map((role) => role.label);
}

export function listSharedTables(
  page = 1,
  pageSize = 20,
): Promise<AutotoolTablesPage> {
  return api<AutotoolTablesPage>(
    `/autotool/tables?page=${page}&page_size=${pageSize}`,
  );
}

export function getPostPreview(
  tableId: number,
  siteColumnId?: number | null,
  pageSize?: number | null,
): Promise<AutotoolPostPreview> {
  const qs = new URLSearchParams();
  if (siteColumnId != null) qs.set("site_column_id", String(siteColumnId));
  if (pageSize != null) qs.set("page_size", String(pageSize));
  const s = qs.toString();
  return api<AutotoolPostPreview>(
    `/autotool/tables/${tableId}/post-preview${s ? `?${s}` : ""}`,
  );
}

// ----- send runs (background, with a progress page) -----

export type AutotoolRunStatus =
  | "queued"
  | "running"
  | "cancelled"
  | "done"
  | "failed";

export type AutotoolItemStatus =
  | "queued"
  | "sending"
  | "sent"
  | "failed"
  | "skipped";

export interface AutotoolRun {
  id: number;
  table_id: number | null;
  table_name: string;
  target_url: string;
  page_size: number;
  status: AutotoolRunStatus;
  total: number;
  sent: number;
  failed: number;
  skipped: number;
  created_at: string;
  started_at: string | null;
  finished_at: string | null;
}

export interface AutotoolRunItem {
  id: number;
  site: string;
  start: number;
  total: number;
  status: AutotoolItemStatus;
  /** the proxy id for this item's site (null until its leader captures it). */
  external_id: string | number | null;
  status_code: number | null;
  detail: string | null;
  response_snippet: string | null;
  elapsed_ms: number | null;
  created_at: string;
}

export interface AutotoolRunsPage {
  items: AutotoolRun[];
  total: number;
  page: number;
  page_size: number;
}

export interface AutotoolRunDetail extends AutotoolRun {
  site_column_id: number | null;
  error: string | null;
  items: AutotoolRunItem[];
  items_total: number;
  items_page: number;
  items_page_size: number;
}

/** Create a background send run and enqueue it. Publishes to live sites — the
 *  caller must confirm first, then redirect to the run's progress page. */
export function createAutotoolRun(
  tableId: number,
  siteColumnId?: number | null,
  pageSize?: number | null,
  acknowledgeAppend = false,
): Promise<AutotoolRunDetail> {
  return api<AutotoolRunDetail>("/autotool/runs", {
    method: "POST",
    body: {
      table_id: tableId,
      site_column_id: siteColumnId ?? null,
      page_size: pageSize ?? null,
      acknowledge_append: acknowledgeAppend,
    },
  });
}

export function listAutotoolRuns(
  page = 1,
  pageSize = 20,
): Promise<AutotoolRunsPage> {
  return api<AutotoolRunsPage>(
    `/autotool/runs?page=${page}&page_size=${pageSize}`,
  );
}

export function getAutotoolRun(
  runId: number,
  page = 1,
  pageSize = 50,
): Promise<AutotoolRunDetail> {
  return api<AutotoolRunDetail>(
    `/autotool/runs/${runId}?page=${page}&page_size=${pageSize}`,
  );
}

export function cancelAutotoolRun(runId: number): Promise<AutotoolRunDetail> {
  return api<AutotoolRunDetail>(`/autotool/runs/${runId}/cancel`, {
    method: "POST",
  });
}

export function retryFailedAutotoolRun(
  runId: number,
): Promise<AutotoolRunDetail> {
  return api<AutotoolRunDetail>(`/autotool/runs/${runId}/retry-failed`, {
    method: "POST",
  });
}
